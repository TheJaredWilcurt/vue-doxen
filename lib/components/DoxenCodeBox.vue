<template>
  <div
    v-bind="applyStyleTokens({ codeBoxContainer: true })"
    aria-label="Code example"
    role="textbox"
    tabindex="0"
    title="Click to copy"
    @click="copy"
    @keydown.enter="copy"
    @keydown.space.prevent.stop="copy"
  >
    <HighlightJS
      v-bind="applyStyleTokens({ codeBox: true })"
      :autodetect="false"
      :code="code"
      :language="language"
      :key="index"
    />
    <div
      v-if="copied"
      v-bind="applyStyleTokens({ codeBoxCopied: copied })"
    >
      Copied
    </div>
  </div>
</template>

<script>
/* eslint-disable import/extensions */
import hljsVuePlugin from '@highlightjs/vue-plugin';
import hljs from 'highlight.js/lib/core';
import javascript from 'highlight.js/lib/languages/javascript';
import xml from 'highlight.js/lib/languages/xml';

import { styleTokens } from '@/helpers/props.js';

import applyStyleTokens from '@/mixins/applyStyleTokensMixin.js';

const JAVASCRIPT = 'javascript';
const XML = 'xml';

hljs.registerLanguage(JAVASCRIPT, javascript);
hljs.registerLanguage(XML, xml);

export default {
  name: 'DoxenCodeBox',
  components: {
    HighlightJS: hljsVuePlugin.component
  },
  mixins: [applyStyleTokens],
  props: {
    code: {
      type: String,
      default: undefined
    },
    styleTokens
  },
  data: function () {
    return {
      copied: false,
      index: 0
    };
  },
  methods: {
    copy: async function () {
      try {
        await navigator.clipboard.writeText(this.code);
        this.copied = true;
        setTimeout(() => {
          this.copied = false;
        }, 2100);
      } catch (error) {
        console.error(error.message);
      }
    }
  },
  computed: {
    language: function () {
      if (this.code.trim().startsWith('<')) {
        return XML;
      } else {
        return JAVASCRIPT;
      }
    }
  },
  created: function () {
    // Force event loop to cycle, then trigger
    // re-applying syntax highlighting
    setTimeout(() => {
      this.index++;
    }, 0);
  }
};
</script>
