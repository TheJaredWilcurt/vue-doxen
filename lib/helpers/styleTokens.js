/**
 * @file Style tokens allow a lot of flexibility in how the Vue-Doxen components can be styled.
 *
 * Style tokens solve the following scenarios:
 *
 * Provide optional basic built in CSS
 * Allow overriding basic built-in classes
 * The CSS is externalized and can be skipped if not needed
 * You can remove all classes from all elements if not needed
 * Allows customizing the class names to whatever you want
 * Allows having many classes applied on a single element (Tailwind)
 * Can use existing CSS class names (Bootstrap)
 * Can use existing CSS component classes (Vuetify)
 */

/**
 * A predictable map of tokens to class names. These classes are not used anywhere, they exist to make styling easier for those writing their own CSs.
 *
 * @type {Object}
 */
export const styleTokensBuiltIn = {
  button: 'doxen-button',
  buttonNotSelected: 'doxen-button-not-selected',
  buttonSelected: 'doxen-button-selected',
  codeBox: 'doxen-code-box',
  codeSwapper: 'doxen-code-swapper',
  codeSwapperFileName: 'doxen-code-swapper-file-name',
  componentDemo: 'doxen-component-demo',
  componentDemoContainer: 'doxen-component-demo-container',
  componentDemoH3: 'doxen-component-demo-h3',
  componentDemoHr: 'doxen-component-demo-hr',
  demoDescription: 'doxen-demo-description',
  demoHeader: 'doxen-demo-header',
  demoHeaderTitle: 'doxen-demo-header-title',
  emitsDocumentationCode: 'doxen-emits-documentation-code',
  emitsDocumentationLi: 'doxen-emits-documentation-li',
  emitsDocumentationSpan: 'doxen-emits-documentation-span',
  emitsDocumentationStrong: 'doxen-emits-documentation-strong',
  emitsDocumentationUl: 'doxen-emits-documentation-ul',
  emitLogClearButton: 'doxen-emit-log-clear-button',
  emitLogHeader: 'doxen-emit-log-header',
  emitLogTable: 'doxen-emit-log-table',
  emitLogTableContainer: 'doxen-emit-log-table-container',
  emitLogTbody: 'doxen-emit-log-tbody',
  emitLogTd: 'doxen-emit-log-td',
  emitLogTh: 'doxen-emit-log-th',
  emitLogThead: 'doxen-emit-log-thead',
  emitLogTr: 'doxen-emit-log-tr',
  formFieldCheckbox: 'doxen-form-field-checkbox',
  formFieldCheckboxContainer: 'doxen-form-field-checkbox-container',
  formFieldCheckboxError: 'doxen-form-field-checkbox-error',
  formFieldCheckboxNameDisabled: 'doxen-form-field-checkbox-name-disabled',
  formFieldCheckboxNameError: 'doxen-form-field-checkbox-name-error',
  formFieldCheckboxNameLabel: 'doxen-form-field-checkbox-name-label',
  formFieldDropdown: 'doxen-form-field-dropdown',
  formFieldDropdownContainer: 'doxen-form-field-dropdown-container',
  formFieldDropdownError: 'doxen-form-field-dropdown-error',
  formFieldDropdownOption: 'doxen-form-field-dropdown-option',
  formFieldError: 'doxen-form-field-error',
  formFieldFieldset: 'doxen-form-field-fieldset',
  formFieldLabel: 'doxen-form-field-label',
  formFieldLegend: 'doxen-form-field-legend',
  formFieldLegendDisabled: 'doxen-form-field-legend-disabled',
  formFieldLegendError: 'doxen-form-field-legend-error',
  formFieldMessage: 'doxen-form-field-message',
  formFieldRadioDial: 'doxen-form-field-radio-dial',
  formFieldRadioDialContainer: 'doxen-form-field-radio-dial-container',
  formFieldRadioDialNameDisabled: 'doxen-form-field-radio-dial-name-disabled',
  formFieldRadioDialNameError: 'doxen-form-field-radio-dial-name-error',
  formFieldRadioDialNameLabel: 'doxen-form-field-radio-dial-name-label',
  formFieldRadioDialsContainer: 'doxen-form-field-radio-dials-container',
  formFieldRequired: 'doxen-form-field-required',
  formFieldTextarea: 'doxen-form-field-textarea',
  formFieldTextareaCode: 'doxen-form-field-textarea-code',
  formFieldTextareaContainer: 'doxen-form-field-textarea-container',
  formFieldTextareaError: 'doxen-form-field-textarea-error',
  formFieldTextField: 'doxen-form-field-text-field',
  formFieldTextFieldContainer: 'doxen-form-field-text-field-container',
  formFieldTextFieldError: 'doxen-form-field-text-field-error',
  propsDocumentationCode: 'doxen-props-documentation-code',
  propsDocumentationEm: 'doxen-props-documentation-em',
  propsDocumentationLi: 'doxen-props-documentation-li',
  propsDocumentationSpan: 'doxen-props-documentation-span',
  propsDocumentationStrong: 'doxen-props-documentation-strong',
  propsDocumentationUl: 'doxen-props-documentation-ul',
  propsPlaygroundForm: 'doxen-props-playground-form',
  sidebarButton: 'doxen-sidebar-button',
  sidebarButtonNotSelected: 'doxen-sidebar-button-not-selected',
  sidebarButtonSelected: 'doxen-sidebar-button-selected',
  sidebarContainer: 'doxen-sidebar-container',
  tabButton: 'doxen-tab-button',
  tabButtonNotSelected: 'doxen-tab-button-not-selected',
  tabButtonSelected: 'doxen-tab-button-selected',
  tabsButtonContainer: 'doxen-tabs-button-container',
  tabsHeader: 'doxen-tabs-header',
  vueDoxen: 'doxen-vue-doxen'
};

/**
 * All the style token keys without any class names applied to them.
 *
 * @type {Object}
 */
export const styleTokensEmpty = {};
for (const key in styleTokensBuiltIn) {
  styleTokensEmpty[key] = '';
}

/**
 * Style tokens using the class names provided by Bootstrap 5.
 *
 * @type {Object}
 */
export const styleTokensBootstrap5 = {
  // Container
  formFieldFieldset: 'px-2 pb-2 m-1 flex-grow-1',
  // Errors
  formFieldCheckboxError: 'text-danger',
  formFieldCheckboxNameError: 'text-danger',
  formFieldDropdownError: 'text-danger',
  formFieldError: 'form-text text-danger',
  formFieldLegendError: 'text-danger',
  formFieldRadioDialNameError: 'text-danger',
  formFieldTextareaError: 'text-danger',
  formFieldTextFieldError: 'text-danger',
  // Labels
  formFieldLegend: 'form-label form-control-sm mb-0 px-0',
  // Inputs
  formFieldTextarea: 'form-control',
  formFieldTextField: 'form-control',
  // Checkbox
  formFieldCheckbox: 'form-check-input',
  formFieldCheckboxContainer: 'form-check',
  formFieldCheckboxNameDisabled: '',
  formFieldCheckboxNameLabel: 'form-check-label',
  // Radio Dials
  formFieldRadioDial: 'form-check-input',
  formFieldRadioDialNameLabel: 'form-check-label form-label form-control-sm',
  formFieldRadioDialContainer: 'form-check',
  // Select
  formFieldDropdown: 'form-select form-select-sm',
  // Message
  formFieldMessage: 'form-text',
  // Buttons
  button: 'btn',
  buttonSelected: 'btn-primary',
  buttonNotSelected: 'btn-outline-primary',
  // Code
  codeSwapperFileName: 'float-end',
  formFieldTextareaCode: 'font-monospace',
  tabsButtonContainer: 'nav nav-tabs nav-item mt-3 mb-0',
  tabButton: 'nav-link me-1',
  tabButtonSelected: 'active border-bottom-0 mb-0',
  // Demo
  propsPlaygroundForm: 'd-flex flex-wrap',
  // Emit log
  emitLogHeader: 'd-flex',
  emitLogClearButton: 'btn btn-primary',
  emitLogTable: 'table table-striped',
  // Sidebar
  sidebarContainer: 'btn-group-vertical mx-3 justify-content-start',
  sidebarButton: 'btn flex-grow-0',
  sidebarButtonNotSelected: 'btn-outline-primary',
  sidebarButtonSelected: 'btn-primary flex-shrink-1'
};
