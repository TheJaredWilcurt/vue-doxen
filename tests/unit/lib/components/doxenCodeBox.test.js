import { styleTokensBuiltIn } from '@/helpers/styleTokens.js';

import DoxenCodeBox from '@/components/DoxenCodeBox.vue';

import testHelpers from '@@/unit/testHelpers.js';

describe('DoxenCodeBox', () => {
  const consoleError = console.error;

  // Props
  const html = '<div attribute="value">\n  Text\n</div>';
  const js = 'const value = 4;\nconsole.log({\n  key: value\n});';
  const code = '<h1>Text</h1>';
  const object = {
    key: 'value',
    fn: () => { return 2; }
  };
  const styleTokens = styleTokensBuiltIn;
  const requiredProps = {};

  beforeEach(() => {
    console.error = vi.fn();
    vi.useFakeTimers();
  });

  afterEach(() => {
    console.error = consoleError;
    vi.useRealTimers();
  });

  const setupWrapper = async (props) => {
    const options = {
      props: {
        ...requiredProps,
        ...props
      }
    };
    const wrapper = await testHelpers.mount(DoxenCodeBox, options);
    return wrapper;
  };

  test('Renders with just required props', async () => {
    const wrapper = await setupWrapper();

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Renders HTML', async () => {
    const props = {
      code: html,
      styleTokens
    };
    const wrapper = await setupWrapper(props);

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Renders JavaScript', async () => {
    const props = {
      code: js,
      styleTokens
    };
    const wrapper = await setupWrapper(props);

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Serialize JS', async () => {
    const props = {
      code: object,
      styleTokens
    };
    const wrapper = await setupWrapper(props);

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Copy disabled', async () => {
    const props = {
      code,
      copy: false,
      styleTokens
    };
    const wrapper = await setupWrapper(props);

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Copy text temporarily displayed after clicking on codebox', async () => {
    const wrapper = await setupWrapper({ code });
    const firstDiv = wrapper.find('[aria-label="Code example"]');
    await firstDiv.trigger('click');

    expect(wrapper.vm.copied)
      .toEqual(true);

    vi.runAllTimers();

    expect(wrapper.vm.copied)
      .toEqual(false);
  });

  test('Copy text temporarily displayed after pressing enter', async () => {
    const wrapper = await setupWrapper({ code });
    const firstDiv = wrapper.find('[aria-label="Code example"]');
    await firstDiv.trigger('keydown.enter');

    expect(wrapper.vm.copied)
      .toEqual(true);

    vi.runAllTimers();

    expect(wrapper.vm.copied)
      .toEqual(false);
  });

  test('Copy text temporarily displayed after pressing space', async () => {
    const wrapper = await setupWrapper({ code });
    const firstDiv = wrapper.find('[aria-label="Code example"]');
    await firstDiv.trigger('keydown.space');

    expect(wrapper.vm.copied)
      .toEqual(true);

    vi.runAllTimers();

    expect(wrapper.vm.copied)
      .toEqual(false);
  });

  describe('Mock clipboard', () => {
    let consoleError;
    let writeText;

    beforeEach(() => {
      consoleError = console.error;
      console.error = vi.fn();

      writeText = navigator.clipboard.writeText;
      navigator.clipboard.writeText = vi.fn(() => {
        throw new Error('Clipboard permission missing.');
      });
    });

    afterEach(() => {
      console.error = consoleError;
      navigator.clipboard.writeText = writeText;
    });

    test('Catch copy error.', async () => {
      const wrapper = await setupWrapper({ code });
      const firstDiv = wrapper.find('[aria-label="Code example"]');
      await firstDiv.trigger('click');

      expect(navigator.clipboard.writeText)
        .toHaveBeenCalledWith(code);

      expect(console.error)
        .toHaveBeenCalledWith('Clipboard permission missing.');
    });
  });
});
