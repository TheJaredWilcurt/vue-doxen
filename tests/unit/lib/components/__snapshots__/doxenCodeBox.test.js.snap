// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DoxenCodeBox > Copy disabled 1`] = `
<div
  aria-label="Code example"
  class="doxen-code-box-container"
  data-applied-style-tokens="codeBoxContainer"
  data-style-tokens="codeBoxContainer"
  role="textbox"
  tabindex="0"
>
  <pre
    class="doxen-code-box"
    data-applied-style-tokens="codeBox"
    data-style-tokens="codeBox"
  ><code
    class="
      hljs
      xml
    "
    tabindex="0"
  ><span class="hljs-tag">&lt;<span class="hljs-name">h1</span>&gt;</span>Text<span class="hljs-tag">&lt;/<span class="hljs-name">h1</span>&gt;</span></code></pre>
  <!-- v-if -->
</div>
`;

exports[`DoxenCodeBox > Renders HTML 1`] = `
<div
  aria-label="Code example"
  class="doxen-code-box-container"
  data-applied-style-tokens="codeBoxContainer"
  data-style-tokens="codeBoxContainer"
  role="textbox"
  tabindex="0"
  title="Click to copy"
>
  <pre
    class="doxen-code-box"
    data-applied-style-tokens="codeBox"
    data-style-tokens="codeBox"
  ><code
    class="
      hljs
      xml
    "
    tabindex="0"
  ><span class="hljs-tag">&lt;<span class="hljs-name">div</span> <span class="hljs-attr">attribute</span>=<span class="hljs-string">"value"</span>&gt;</span>
  Text
<span class="hljs-tag">&lt;/<span class="hljs-name">div</span>&gt;</span></code></pre>
  <!-- v-if -->
</div>
`;

exports[`DoxenCodeBox > Renders JavaScript 1`] = `
<div
  aria-label="Code example"
  class="doxen-code-box-container"
  data-applied-style-tokens="codeBoxContainer"
  data-style-tokens="codeBoxContainer"
  role="textbox"
  tabindex="0"
  title="Click to copy"
>
  <pre
    class="doxen-code-box"
    data-applied-style-tokens="codeBox"
    data-style-tokens="codeBox"
  ><code
    class="
      hljs
      javascript
    "
    tabindex="0"
  ><span class="hljs-keyword">const</span> value = <span class="hljs-number">4</span>;
<span class="
    hljs-variable
    language_
  ">console</span>.<span class="
    function_
    hljs-title
  ">log</span>({
  <span class="hljs-attr">key</span>: value
});</code></pre>
  <!-- v-if -->
</div>
`;

exports[`DoxenCodeBox > Renders with just required props 1`] = `
<div
  aria-label="Code example"
  class=""
  data-applied-style-tokens="codeBoxContainer"
  data-style-tokens="codeBoxContainer"
  role="textbox"
  tabindex="0"
  title="Click to copy"
>
  <pre
    class=""
    data-applied-style-tokens="codeBox"
    data-style-tokens="codeBox"
  ><code
    class="
      hljs
      javascript
    "
    tabindex="0"
  ></code></pre>
  <!-- v-if -->
</div>
`;

exports[`DoxenCodeBox > Serialize JS 1`] = `
<div
  aria-label="Code example"
  class="doxen-code-box-container"
  data-applied-style-tokens="codeBoxContainer"
  data-style-tokens="codeBoxContainer"
  role="textbox"
  tabindex="0"
  title="Click to copy"
>
  <pre
    class="doxen-code-box"
    data-applied-style-tokens="codeBox"
    data-style-tokens="codeBox"
  ><code
    class="
      hljs
      javascript
    "
    tabindex="0"
  >{
  <span class="hljs-attr">key</span>: <span class="hljs-string">'value'</span>,
  <span class="hljs-attr">fn</span>: <span class="hljs-function">() =&gt;</span> { <span class="hljs-keyword">return</span> <span class="hljs-number">2</span>; }
}</code></pre>
  <!-- v-if -->
</div>
`;
